README.md
# django-enumfield

Provides an enumeration Django model field (using `IntegerField`) with reusable enums and transition validation.

[![Build Status](https://github.com/5monkeys/django-enumfield/workflows/Test/badge.svg)](https://github.com/5monkeys/django-enumfield/actions)
[![PyPi Version](https://img.shields.io/pypi/v/django-enumfield.svg)](https://pypi.python.org/pypi/django-enumfield)
[![License](https://img.shields.io/pypi/l/django-enumfield.svg)](https://pypi.python.org/pypi/django-enumfield)
[![Python Versions](https://img.shields.io/pypi/pyversions/django-enumfield.svg)](https://pypi.python.org/pypi/django-enumfield)
[![Wheel](https://img.shields.io/pypi/wheel/django-enumfield.svg)](https://pypi.python.org/pypi/django-enumfield)
![Coveralls github](https://img.shields.io/coveralls/github/5monkeys/django-enumfield)

Installation
------------

Currently, [we test](https://github.com/5monkeys/django-enumfield/actions) Django versions 2.2-4.1 and Python versions 3.7-3.11.

Install `django-enumfield` in your Python environment:

```sh
$ pip install django-enumfield
```

**Upgrading from django-enumfield 1.x?** [See the migration guide](docs/migrate-to-20.md)

For use with Django versions prior to 1.8 use version
[`1.2.1`](https://github.com/5monkeys/django-enumfield/tree/1.2.1)

For use with Django versions prior to 1.11 use version
[`1.5`](https://github.com/5monkeys/django-enumfield/tree/1.5)

Usage
-----

Create an `Enum`-class and pass it as first argument to the Django model `EnumField`.

```python
from django.db import models
from django_enumfield import enum


class BeerStyle(enum.Enum):
    LAGER = 0
    STOUT = 1
    WEISSBIER = 2


class Beer(models.Model):
    style = enum.EnumField(BeerStyle, default=BeerStyle.LAGER)


# Use .get to get enum values from either name or ints
print(BeerStyle.get("LAGER"))  # <BeerStyle.LAGER: 0>
print(BeerStyle.get(1))  # <BeerStyle.STOUT: 1>
print(BeerStyle.get(BeerStyle.WEISSBIER))  # <BeerStyle.WEISSBIER: 2>

# It's also possible to use the normal enum way to get the value
print(BeerStyle(1))  # <BeerStyle.STOUT: 1>
print(BeerStyle["LAGER"])  # <BeerStyle.LAGER: 0>

# The enum value has easy access to their value and name
print(BeerStyle.LAGER.value)  # 0
print(BeerStyle.LAGER.name)  # "LAGER"
```

For more information about Python 3 enums
(which our `Enum` inherits, `IntEnum` to be specific)
checkout the [docs](https://docs.python.org/3/library/enum.html).


### Setting the default value

You can also set default value on your enum class using `__default__`
attribute

```python
from django.db import models
from django_enumfield import enum


class BeerStyle(enum.Enum):
    LAGER = 0
    STOUT = 1
    WEISSBIER = 2

    __default__ = LAGER


class BeerStyleNoDefault(enum.Enum):
    LAGER = 0


class Beer(models.Model):
    style_default_lager = enum.EnumField(BeerStyle)
    style_default_stout = enum.EnumField(BeerStyle, default=BeerStyle.STOUT)
    style_default_null = enum.EnumField(BeerStyleNoDefault, null=True, blank=True)


# When you set __default__ attribute, you can access default value via
# `.default()` method of your enum class
assert BeerStyle.default() == BeerStyle.LAGER

beer = Beer.objects.create()
assert beer.style_default_larger == BeerStyle.LAGER
assert beer.style_default_stout == BeerStyle.STOUT
assert beer.style_default_null is None
```

### Labels

You can use your own labels for `Enum` items

```python
from django.utils.translation import gettext_lazy
from django_enumfield import enum


class Animals(enum.Enum):
    CAT = 1
    DOG = 2
    SHARK = 3

    __labels__ = {
        CAT: gettext_lazy("Cat"),
        DOG: gettext_lazy("Dog"),
    }


print(Animals.CAT.label)  # "Cat"
print(Animals.SHARK.label)  # "SHARK"

# There's also classmethods for getting the label
print(Animals.get_label(2))  # "Dog"
print(Animals.get_label("DOG"))  # "Dog"
```

### Validate transitions

The `Enum`-class provides the possibility to use transition validation.

```python
from django.db import models
from django_enumfield import enum
from django_enumfield.exceptions import InvalidStatusOperationError


class PersonStatus(enum.Enum):
    ALIVE = 1
    DEAD = 2
    REANIMATED = 3

    __transitions__ = {
        DEAD: (ALIVE,),  # Can go from ALIVE to DEAD
        REANIMATED: (DEAD,)  # Can go from DEAD to REANIMATED
    }


class Person(models.Model):
    status = enum.EnumField(PersonStatus)

# These transitions state that a PersonStatus can only go to DEAD from ALIVE and to REANIMATED from DEAD.
person = Person.objects.create(status=PersonStatus.ALIVE)
try:
    person.status = PersonStatus.REANIMATED
except InvalidStatusOperationError:
    print("Person status can not go from ALIVE to REANIMATED")
else:
    # All good
    person.save()
```

### In forms

The `Enum`-class can also be used without the `EnumField`. This is very useful in Django form `ChoiceField`s.

```python
from django import forms
from django_enumfield import enum
from django_enumfield.forms.fields import EnumChoiceField


class GenderEnum(enum.Enum):
    MALE = 1
    FEMALE = 2

    __labels__ = {
        MALE: "Male",
        FEMALE: "Female",
    }


class PersonForm(forms.Form):
    gender = EnumChoiceField(GenderEnum)
```

Rendering `PersonForm` in a template will generate a select-box with "Male" and "Female" as option labels for the gender field.


Local Development Environment
-----------------------------

Make sure black and isort is installed in your env with `pip install -e .[dev]`.

Before committing run `make format` to apply black and isort to all files.


CHANGELOG.md
# Changelog

## [unreleased]

## [3.1.0]

- Support Python 3.11 and Django 4.1 (by [@vitaliyf](https://github.com/vitaliyf))

## [3.0.0]

- Move CI to GitHub Actions
- Dropped support for Python < 3.7
- Dropped support for Django < 2.2
- Added support for Django 3.2
- Added support for Python 3.10
- Added support for Django 4.0

## [2.0.2]

- Added Django 3.1 support. (Pull #63)

## [2.0.1]

- Fixed get_FIELD_display to handle `None`. (Pull #59)

## [2.0.0]

**Many breaking changes this release.**

- The ``enumfield.enum.Enum`` class is now a subclass of the native `IntEnum`
shipped with Python 3.4 (uses the ``enum34`` package on previous versions of Python)
- Renamed `labels` to `__labels__`
- Renamed `_transitions` to `__transitions__`
- Added aliases for the classmethods `Enum.name()` as `Enum.get_name()` and
`Enum.label()` as `Enum.get_label()`.  Access the old way
(`Enum.name()` and `Enum.label()`) is still supported though, but the new names
are easier to be discovered by IDEs for example.
- `Enum.get_label()` and `Enum.get_name()` now return None if the enum value was
not found instead of raising `AttributeError`
- `EnumField` does not automatically set a default which is the first enum value anymore.
Use `Enum.__default__ = VALUE` or pass it explicitly to `EnumField`
- Converted README.rst to markdown (README.md)
- Added Django 2.2 support
- Added Django 3.0b1 support
- Dropped support for Django < 1.11
- Added limited mypy support

## [1.5.0]

- Added Django 2.1 support
- Added Python 3.7 support
- Dropped Python 3.3 support

## [1.4.0]

- Added Django 1.11 support (from [#43](https://github.com/5monkeys/django-enumfield/pull/43))
- Added Python 3.6 support
- Dropped support for Django < 1.8


django_enumfield/enum.py
from __future__ import absolute_import

import logging
import enum
from typing import (
    Any,
    List,
    Optional,
    Sequence,
    Tuple,
    TypeVar,
    Union,
    cast,
    Mapping,
    TYPE_CHECKING,
)
from django.utils.encoding import force_str

if TYPE_CHECKING:
    from django.utils.functional import _StrOrPromise as StrOrPromise

try:
    from django.utils.functional import classproperty  # type: ignore
except ImportError:
    # Pre-Django 3.1
    from django.utils.decorators import classproperty

from django_enumfield.db.fields import EnumField

__all__ = ("Enum", "EnumField")

logger = logging.getLogger(__name__)
RAISE = object()


class BlankEnum(enum.Enum):
    BLANK = ""

    @property
    def label(self):
        return ""


def classdispatcher(class_method):
    class _classdispatcher(object):
        def __init__(self, method=None):
            self.fget = method

        def __get__(self, instance, cls=None):
            if instance is None:
                return getattr(cls, class_method)
            return self.fget(instance)

    return _classdispatcher


Default = TypeVar("Default")
T = TypeVar("T", bound="Enum")


class Enum(enum.IntEnum):
    """A container for holding and restoring enum values"""

    __labels__ = {}  # type: Mapping[int, StrOrPromise]
    __default__ = None  # type: Optional[int]
    __transitions__ = {}  # type: Mapping[int, Sequence[int]]

    def __str__(self):
        return self.label

    @classdispatcher("get_name")
    def name(self):
        # type: () -> str
        return self._name_

    @classdispatcher("get_label")
    def label(self):
        # type: () -> str
        """Get human readable label for the matching Enum.Value.
        :return: label for value
        :rtype: str
        """
        labels = self.__class__.__labels__
        return force_str(labels.get(self.value, self.name))

    @classproperty  # type: ignore[arg-type]
    def do_not_call_in_templates(cls):
        # type: () -> bool
        # Fix for Django templates so that any lookups of enums won't fail
        # More info: https://stackoverflow.com/questions/35953132/how-to-access-enum-types-in-django-templates  # noqa: E501
        return True

    @classproperty  # type: ignore[arg-type]
    def values(cls):
        # type: () -> Mapping[int, Enum]
        return {member.value: member for member in cls}  # type: ignore[attr-defined]

    def deconstruct(self):
        """
        See "Adding a deconstruct() method" in
        https://docs.djangoproject.com/en/1.8/topics/migrations/
        """
        c = self.__class__
        path = "{}.{}".format(c.__module__, c.__name__)
        return path, [self.value], {}

    @classmethod
    def items(cls):
        # type: () -> List[Tuple[str, int]]
        """
        :return: List of tuples consisting of every enum value in the form
            [('NAME', value), ...]
        """
        items = [(member.name, member.value) for member in cls]
        return sorted(items, key=lambda x: x[1])

    @classmethod
    def choices(cls, blank=False):
        # type: (bool) -> List[Tuple[Union[int, str], enum.Enum]]
        """Choices for Enum
        :return: List of tuples (<value>, <member>)
        """
        choices = sorted(
            [(member.value, member) for member in cls], key=lambda x: x[0]
        )  # type: List[Tuple[Union[str, int], enum.Enum]]
        if blank:
            choices.insert(0, (BlankEnum.BLANK.value, BlankEnum.BLANK))
        return choices

    @classmethod
    def default(cls):
        # type: () -> Optional[Enum]
        """Default Enum value. Set default value to `__default__` attribute
        of your enum class or override this method if you need another
        default value.
        Usage:
            IntegerField(choices=my_enum.choices(), default=my_enum.default(), ...
        :return Default value, if set.
        """
        if cls.__default__ is not None:
            return cast(Enum, cls(cls.__default__))
        return None

    @classmethod
    def field(cls, **kwargs):
        # type: (Any) -> EnumField
        """A shortcut for field declaration
        Usage:
            class MyModelStatuses(Enum):
                UNKNOWN = 0

            class MyModel(Model):
                status = MyModelStatuses.field()

        :param kwargs: Arguments passed in EnumField.__init__()
        :rtype: EnumField
        """
        return EnumField(cls, **kwargs)

    @classmethod
    def get(
        cls,
        name_or_numeric,  # type: Union[str, int, T]
        default=None,  # type: Optional[Default]
    ):
        # type: (...) -> Union[Enum, Optional[Default]]
        """Get Enum.Value object matching the value argument.
        :param name_or_numeric: Integer value or attribute name
        :param default: The default to return if the value passed is not
            a valid enum value
        """
        if isinstance(name_or_numeric, cls):
            return name_or_numeric

        if isinstance(name_or_numeric, int):
            try:
                return cls(name_or_numeric)
            except ValueError:
                pass
        elif isinstance(name_or_numeric, str):
            try:
                return cls[name_or_numeric]
            except KeyError:
                pass

        return default

    @classmethod
    def get_name(cls, name_or_numeric):
        # type: (Union[str, int, T]) -> Optional[str]
        """Get Enum.Value name matching the value argument.
        :param name_or_numeric: Integer value or attribute name
        :return: The name or None if not found
        """
        value = cls.get(name_or_numeric)
        if value is not None:
            return value.name
        return None

    @classmethod
    def get_label(cls, name_or_numeric):
        # type: (Union[str, int, Enum]) -> Optional[str]
        """Get Enum.Value label matching the value argument.
        :param name_or_numeric: Integer value or attribute name
        :return: The label or None if not found
        """
        value = cls.get(name_or_numeric)
        if value is not None:
            return value.label
        return None

    @classmethod
    def is_valid_transition(cls, from_value, to_value):
        # type: (Union[int, Enum], Union[int, Enum]) -> bool
        """Will check if to_value is a valid transition from from_value.
        Returns true if it is a valid transition.

        :param from_value: Start transition point
        :param to_value: End transition point
        :return: Success flag
        """
        if isinstance(from_value, cls):
            from_value = from_value.value
        if isinstance(to_value, cls):
            to_value = to_value.value

        return (
            from_value == to_value
            or not cls.__transitions__
            or (from_value in cls.transition_origins(to_value))
        )

    @classmethod
    def transition_origins(cls, to_value):
        # type: (Union[int, T]) -> Sequence[int]
        """Returns all values the to_value can make a transition from.
        :param to_value End transition point
        """
        if isinstance(to_value, cls):
            to_value = to_value.value

        return cls.__transitions__.get(to_value, [])
